src/Compression/CompressedReadBufferBase.cpp

```diff
@@ -142,8 +136,8 @@ size_t CompressedReadBufferBase::readCompressedData(size_t & size_decompressed,
         }
     }
 
-    size_compressed_without_checksum = ICompressionCodec::readCompressedBlockSize(compressed_header);
-    size_decompressed = ICompressionCodec::readDecompressedBlockSize(compressed_header);
+    size_compressed_without_checksum = ICompressionCodec::readCompressedBlockSize(compressed_buffer);
+    size_decompressed = ICompressionCodec::readDecompressedBlockSize(compressed_buffer);
```

Comment:
I think there is a little glitch, namely that in l. 155, the internal statistics `ProfileEvents::increment()` are now potentially updated twice,
- once via CompressedReadBufferFromFile.cpp:109 (`readCompressedDataBlockForAsynchronous`) and
- once via CompressedReadBufferFromFile.cpp:123 (`readCompressedData`). 

Maybe add a bool parameter to `readHeaderAndGetCodecAndSize()` which is true in the former case and false in the latter case?

```
if (updateReadCompressedBytes)
    ProfileEvents::increment(ProfileEvents::ReadCompressedBytes, size_compressed_without_checksum + sizeof(Checksum));
```

Reply:
I removed "ProfileEvents::increment" from readHeaderAndGetCodecAndSize. And copy it respectively to the end of readCompressedDataBlockForAsynchronous and readCompressedData which means it only be called when "compressed_in" shifting offset successfully.